src/cdk/schematics/update-tool/index.ts

```diff
@@ -6,152 +6,140 @@
+ * accepts a TypeScript program and a context that is provided to all migrations. The
+ * context is usually not used by migrations, but in some cases migrations rely on
+ * specifics from the tool that performs the update (e.g. the Angular CLI). In those cases,
+ * the context can provide the necessary specifics to the migrations in a type-safe way.
+ */
+export class UpdateProject<Context> {
+  private readonly _typeChecker: ts.TypeChecker = this._program.getTypeChecker();
+
+  constructor(private _context: Context,
+              private _program: ts.Program,
+              private _fileSystem: FileSystem,
+              private _analyzedFiles: Set<string> = new Set(),
+              private _logger: UpdateLogger = defaultLogger) {}
+
+  /**
+   * Migrates the project to the specified target version.
+   * @param migrationTypes Migrations that should be run.
+   * @param target Version the project should be updated to.
+   * @param data Upgrade data that is passed to all migration rules.
+   * @param additionalStylesheetPaths Additional stylesheets that should be migrated, if not
+   *   referenced in an Angular component. This is helpful for global stylesheets in a project.
+   */
+  migrate<Data>(migrationTypes: MigrationCtor<Data, Context>[], target: TargetVersion, data: Data,
+      additionalStylesheetPaths?: string[]): {hasFailures: boolean} {
+    const migrations: Migration<Data, Context>[] = [];
+
+    // Create instances of all specified migrations.
+    for (const ctor of migrationTypes) {
+      const instance = new ctor(this._program, this._typeChecker, target, this._context,
+          data, this._fileSystem, this._logger);
+      instance.init();
+      if (instance.enabled) {
+        migrations.push(instance);
+      }
     }
-  });
-
-  // In some applications, developers will have global stylesheets which are not specified in any
-  // Angular component. Therefore we glob up all CSS and SCSS files outside of node_modules and
-  // dist. The files will be read by the individual stylesheet rules and checked.
-  // TODO: rework this to collect external/global stylesheets from the workspace config. COMP-280.
-  globSync(
-      '!(node_modules|dist)/**/*.+(css|scss)', {absolute: true, cwd: projectFsPath, nodir: true})
-      .filter(filePath => !resourceCollector.resolvedStylesheets.some(s => s.filePath === filePath))
-      .forEach(filePath => {
-        const stylesheet = resourceCollector.resolveExternalStylesheet(filePath, null);
-        const relativePath = getProjectRelativePath(filePath);
-        // do not visit stylesheets which have been referenced from a component.
-        if (!analyzedFiles.has(relativePath)) {
-          rules.forEach(r => r.visitStylesheet(stylesheet));
-        }
-      });
 
-  // Call the "postAnalysis" method for each migration rule.
-  rules.forEach(r => r.postAnalysis());
+    const sourceFiles = this._program.getSourceFiles().filter(
```

Comment:
I would consider breaking this function up into smaller functions, or maybe just adding some signpost comments. Something like 

```ts
// Collect all of the TypeScript source files we want to migrate,
// which excludes `d.ts` files and external dependencies.
const sourceFiles = this._program.getSourceFiles().filter(...);

// Collect all of the HTML and CSS resources we want to migrate.
const resourceCollector = new ComponentResourceCollector(this._typeChecker, this._fileSystem);
...

// Collect HTML and CSS resources for each TypeScript source file
sourceFiles.forEach(...)

...
```

Reply:
Moved the migration creation into a separate function and added comments.